### ISSUE TITLE
Making ByteString an abstract class

### ISSUE THREAD

**Issue by kevincianfarini** [square/okio#736]
I've been looking into how I can address #668 in my free time.

After realizing that `ByteString` is a concrete class and not an abstract class, I got to wondering how `SegmentedByteString` works in the first place. Turns out, every operation done on `SegmentedByteString` performs a copy of all of the data contained within to a single array, and then converts that to a normal `ByteString`. 

The restrictions on subclassing `ByteString` make it close to impossible to address `ByteString.slice` and also raise questions as to why `SegmentedByteString` requires copies. 

Making `ByteString` abstract would address all of these problem. I have a proof of concept available in my personal project Corbit in the [corbit-binary](https://github.com/kevincianfarini/corbit/blob/master/corbit-binary/src/commonMain/kotlin/com/corbit/binary) module. The key to flexibility here is operators like `ByteString.get` would be abstract. 

The idea here would be to have all concrete implementations of `ByteString` internal to okio (unless necessary) and only expose the type `ByteString`. This allows:

* zero copy functionality on `SegmentedByteString`
* `ByteArrayByteString`
* `ByteString.slice` -> `ByteStringSlice`

**I have absolutely no idea what affect this has on binary/api compatibility, but would love to have this conversation about feasibility**

**Comment by swankjesse** [comment]
I do not want a zero copy slice feature. It creates a potential for memory leaks! And I don't believe the benefits of zero copy slice are particularly high... you need to be slicing a large string many times to get the benefit, and I don’t expect that to be common.

**Comment by kevincianfarini** [comment]
Yeah, memory leaks have been on my mind as well while working with slice. 

That being said, should we forbid the consumers of okio from writing their own slice extension on ByteString? 

Making ByteString an abstract class gets rid of copying in SegmentedByteSteing and also allows a bit of flexibility consumers who know they need it.

**Comment by JakeWharton** [comment]
I am reading from a tens-of-megabyte bytestring in kilobyte to megabyte chunks many times.

I'm not really convinced by memory leak arguments here. That's at worst a naming problem but not a reason to discard the feature seeing as you can memory leak just about anything without care.

**Comment by swankjesse** [comment]
They changed their minds on OpenJDK’s string ’cause it caused too much trouble.
https://bugs.openjdk.java.net/browse/JDK-6924259

If you really really wanted this you could create a different type, `ByteStringSlice` that holds a `ByteString` and offsets. A different type is useful because then it’s clear that the slices have different performance characteristics from regular ByteStrings.

**Comment by kevincianfarini** [comment]
Creating a `ByteStringSlice` is the proposed solution I mentioned above. 

However, i believe it doesn't really make sense for it to be a completely disjoint type from `ByteString`. At best, all of the convenience functionality would need to be reimplemented on slice versus having it implemented on an abstract class or extensions. 

Something that concerns me more than the validity of slice is that `SegmentedByteString` makes copies for several of it's operations. Doesn't this negate any performance gain of having a segment oriented byte string in the first place?

Please correct me if I'm wrong. If we can come to an agreement on this, I'm interested in doing the work that's needed.

**Comment by JakeWharton** [comment]
The JDK changed `substring` which was the only mechanism of taking a subsection of a string. In that world doing a copy makes sense. But there is no `slice` method on string, and they probably should have introduced one as part of that change to not harm people who knew the implications of such an operation.

`ByteString` already has a substring with the safe copy semantics. We could even call it unsafeSlice or any sufficiently scary name.

I suspect the new incubating Vector API in the JDK will have a slice operation available.

**Comment by kevincianfarini** [comment]
I'm willing to make a PoC implementation. Is that of any interest?

**Comment by swankjesse** [comment]
I still think this is not solving a real performance problem. If you have a hot spot in your application that this would improve, tell me about it.

**Comment by kevincianfarini** [comment]
There's an issue opened on the backlog, #688, for caching the resulting ByteString in `SegmentedByteString`. By making `ByteString` an abstract class, the `SegmentedByteString` wouldn't need to create and cache an unsegmented instance at all.

The only function on this class which would need to be abstract would be 

```kotlin
abstract operator fun get(index: Int): Byte
```

With this, the implementation of `get` could use the `segments` and `directory` to get the data directly without having to make a copy.
